### PR TITLE
Compiler: refactor sourcemap generation, use Index sourcemap

### LINKED ISSUE #1446: [FEATURE REQUEST] Use source-map sections in js_of_ocaml link command.
Support sourcemap sections fields so that we no longer need to merge sourcemaps mappings in `js_of_ocaml link`
```
Index map: supporting post processing
To support concatenating generated code and other common post processing, an alternate representation of a map is supported:
{
"version" : 3,
"file": “app.js”,
"sections": [
 { "offset": {"line":0, "column":0}, "url": “url_for_part1.map” }
 { "offset": {"line":100, "column":10}, "map": 
   {
     "version" : 3,
     "file": “section.js”,
     "sources": ["foo.js", "bar.js"],
     "names": ["src", "maps", "are", "fun"],
     "mappings": "AAAA,E;;ABCDE;"
   }
 }
],
}


The index map follow the form of the standard map
Line 1: The entire file is an JSON object.
Line 2: The version field. See the description of the standard map.
Line 3: The name field. See the description of the standard map.
Line 4: The sections field.

The “sections” field is an array of JSON objects that itself has two fields “offset” and a source map reference.  “offset” is an object with two fields, “line” and “column”, that represent the offset into generated code that the referenced source map represents.
The other field must be either “url” or “map”. A “url” entry must be a URL where a source map can be found for this section and the url is resolved in the same way as the “sources” fields in the standard map. A “map” entry must be an embedded complete source map object.  An embedded map does not inherit any values from the containing index map.

The sections must be sorted by starting position and the represented sections may not overlap.

```

### PR DESCRIPTION
**Pull request by hhugo**
This PR leverages the recent support for index sourcemap to make sourcemap generation faster.
We no longer decode or edit mappings.

When compiling cma to js. We emit one sourcemap section per compilation unit. This allows to copy one compilation unit with its mappings without having to decode/change its mapping.

Perf
===
Compiling js_of_ocaml with itself (`dune build compiler/bin-js_of_ocaml/js_of_ocaml.bc.js`).
The final link time goes from 660ms on master to 440ms with this PR

Building the toplevel in toplevel/examples/lwt_toplevel/,
The final link time goes from 1350ms (620 ms processing sourcemap) on master to 730ms (90ms processing sourcemap) with this PR
  

X-link
====
build on #1714 
replace #1617 
fix #1446

### CODE REVIEW

**Comment by hhugo** [comment]
@OlivierNicole, do you want to take a look at this ?

File: compiler/lib/link_js.ml [hunk 1]
```diff
@@ -447,22 +449,48 @@ let link ~output ~linkall ~mklib ~toplevel ~files ~resolve_sourcemap_url ~source
...
+                    sm.Source_map.Index.sections
+                    ~f:(fun ({ gen_line; gen_column }, `Map sm) ->
+                      ( gen_line + Source_map.Mappings.first_line sm.mappings
+                      , gen_line + Source_map.Mappings.number_of_lines sm.mappings
+                      , gen_line
+                      , gen_column
+                      , sm ))
+            in
+            List.concat_map reloc ~f:(function
+                | `Drop _ -> []
```
Comment (hhugo) [inline compiler/lib/link_js.ml:473]: This part needs to be reworked

File: compiler/lib/link_js.ml [hunk 2]
```diff
@@ -447,22 +449,48 @@ let link ~output ~linkall ~mklib ~toplevel ~files ~resolve_sourcemap_url ~source
...
+                | `Copy (src, dst, len) ->
+                    List.filter_map sm ~f:(fun (first, last, gen_line, gen_column, sm) ->
+                        if first > src + len || last < src
```
Comment (hhugo) [inline compiler/lib/link_js.ml:476]: This needs to be adjusted

File: compiler/tests-sourcemap/dump_sourcemap.ml [hunk 3]
```diff
@@ -74,12 +74,12 @@ let print_mapping lines ?(line_offset = 0) (sm : Source_map.Standard.t) =
                 ori_line
                 ori_col
                 gen_col
-                (mark gen_col lines.(gen_line - 1 + line_offset))
```
Comment (hhugo) [inline compiler/tests-sourcemap/dump_sourcemap.ml:77]: This diff is suspicious

Comment (hhugo) [reply compiler/tests-sourcemap/dump_sourcemap.ml]: fixed

File: compiler/lib/link_js.ml (hunk 2 above)
Comment (hhugo) [reply compiler/lib/link_js.ml]: added a comment

File: compiler/lib/link_js.ml (hunk 1 above)
Comment (hhugo) [reply compiler/lib/link_js.ml]: done

**Comment by hhugo** [comment]
@rickyvetter, would you be able to test this PR ?

**Comment by OlivierNicole** [comment]
What happens to lines that can be added, such as here? Don’t they require editing the mappings?

https://github.com/ocsigen/js_of_ocaml/blob/81aec6da0250660783a4e3ca8ab1208f8aace610/compiler/lib/link_js.ml#L123

**Comment by OlivierNicole** [comment]
Ok I understand, in the end, it’s fine because it never results in adding lines in the middle of a file, but this mechanism (that sometimes add a `//# <file> <line>` comment when changing files in `Line_writer.write`) seems brittle and makes the code unclear. Could we do this outside of `Line_writer.write`?

**Comment by hhugo** [comment]
> Ok I understand, in the end, it’s fine because it never results in adding lines in the middle of a file, but this mechanism (that sometimes add a `//# <file> <line>` comment when changing files in `Line_writer.write`) seems brittle and makes the code unclear. Could we do this outside of `Line_writer.write`?

I've just simplified the logic a bit.

**Comment by rickyvetter** [comment]
I'm seeing similar improvements to as #1617. I ran across 6 of our bigger apps and they all averaged 45% reduction in linking time.

File: compiler/bin-js_of_ocaml/compile.ml [hunk 4]
```diff
@@ -65,6 +74,52 @@ let output_gen ~standalone ~custom_header ~build_info ~source_map output_file f
...
+  match Builtins.find file with
+  | Some f -> Some (Source_map.Source_content.create (Builtins.File.content f))
+  | None ->
+      if Sys.file_exists file && not (Sys.is_directory file)
+      then
+        let content = Fs.read_file file in
+        Some (Source_map.Source_content.create content)
+      else None
+
+let build_source_map_with_offset
```
Comment (OlivierNicole) [inline compiler/bin-js_of_ocaml/compile.ml:87]: It’s particularly obscure what this function does (apparently, turning a `Source_map.info` into a `Source_map.Standard.t` while ignoring the initial unmapped lines and using filenames from an existing `Source_map.Standard.t`…). A comment would save many minutes of code reading.

File: compiler/bin-js_of_ocaml/compile.ml [hunk 5]
```diff
@@ -65,6 +74,52 @@ let output_gen ~standalone ~custom_header ~build_info ~source_map output_file f
...
+    List.map sources ~f:(fun filename ->
+        match Builtins.find filename with
+        | None -> filename
+        | Some _ -> Filename.concat "/builtin" filename)
+  in
+  let line_offset, mappings = Source_map.Mappings.encode_with_offset mappings in
+  ( line_offset
+  , { (sm : Source_map.Standard.t) with sources; sources_content; names; mappings } )
+
+let build_source_map_many base l =
```
Comment (OlivierNicole) [inline compiler/bin-js_of_ocaml/compile.ml:105]: It’s not much clearer what this one does. In fact, all the code here seems to rest on what `Driver.f` does with sourcemaps, which is undocumented and takes time to figure out as it is spread across so many files. Couldn’t we document `Driver.f`?

File: compiler/lib/source_map.mli [hunk 6]
```diff
@@ -57,10 +59,16 @@ module Mappings : sig
...
       {!val:of_string} and {!val:to_string} are inverse functions. *)
 
-  val decode : t -> map list
+  val decode : t -> decoded
   (** Parse the mappings. *)
 
-  val encode : map list -> t
+  val encode : decoded -> t
+
+  val encode_with_offset : decoded -> int * t
```
Comment (OlivierNicole) [inline compiler/lib/source_map.mli:67]: We should really document how this function differs from `encode` (IIUC it discards the first unmapped lines and returns their number).

**Review by OlivierNicole — APPROVED** [review]
This looks good to me, except for the fact that the sourcemap handling in `Compile` is hard to understand and review, and I think it would really help to add three documentation comments at the right places.

File: compiler/lib/source_map.mli (hunk 6 above)
Comment (hhugo) [reply compiler/lib/source_map.mli]: Done some more refactoring/renaming/documenting

File: compiler/bin-js_of_ocaml/compile.ml (hunk 5 above)
Comment (hhugo) [reply compiler/bin-js_of_ocaml/compile.ml]: What make you think it depends on what `Driver.f` does ? `Driver.f` just return a `Source_map.info`.  There is no much magic there.

File: compiler/bin-js_of_ocaml/compile.ml (hunk 4 above)
Comment (hhugo) [reply compiler/bin-js_of_ocaml/compile.ml]: I've done some refactoring and renaming. I did not add comment yet but it should be much easier to read.

File: compiler/lib/link_js.ml [hunk 7]
```diff
@@ -447,22 +451,62 @@ let link ~output ~linkall ~mklib ~toplevel ~files ~resolve_sourcemap_url ~source
...
+                               sourcemap, but we could in theory filter out part of it. *)
+                            assert (src <= first && last <= src + len);
+                            Some (first, last, gen_line + dst - src, gen_column, sm))))
+            in
+            (* Make sure dropped sections are not overlapping selected sourcemap. *)
+            List.iter reloc ~f:(function
+                | `Copy _ -> ()
+                | `Drop (src, len) ->
+                    List.iter maps ~f:(fun (first, last, _, _, _) ->
+                        if first > src + len || last < src then () else assert false));
```
[the remark this answers is not in the document]
Comment (hhugo) [reply compiler/lib/link_js.ml:490]: no lines are added or removed in the middle of lines covered by a single `Sourcemap.Standard.t`

**Comment by hhugo** [comment]
@OlivierNicole, may I request another round of review ?

File: compiler/bin-js_of_ocaml/compile.ml (hunk 5 above)
Comment (OlivierNicole) [reply compiler/bin-js_of_ocaml/compile.ml]: My bad, I thought there was some accumulation going on in the returned source map but I misread.

**Review by OlivierNicole — COMMENTED** [review]
Looks good to me

**Pull request merged**